### ISSUE TITLE
Transitive path dependencies with different versions cause build error

### ISSUE THREAD

**Issue by bennofs** [rust-lang/cargo#2932]
If you `app` that depends both on `lib-0.1.0` and on `bar`, which in turn depends on `lib-0.2.0`, and both  `app` and `bar` depend on `lib` via a path dependency, then the build will fail with cargo master. This is a regression introduced by commit 0863469c8bb23423b8674c3df857a80eb377348b, since now both rlibs for `lib-0.1.0` and `lib-0.2.0` will share the same name.

**Comment by bennofs** [comment]
An artifical project that causes this issue can be found here: https://github.com/bennofs/repros/tree/master/cargo/%232932

Example `cargo build` output:

```shell
$ /data/code/cargo/target/debug/cargo build
   Compiling lib v0.2.0 (file:///data/code/repros/cargo/%232932/lib-v2)
   Compiling bar v0.1.0 (file:///data/code/repros/cargo/%232932/bar)
   Compiling app v0.1.0 (file:///data/code/repros/cargo/%232932)
src/main.rs:5:5: 5:14 error: unresolved name `lib::foo1` [E0425]
src/main.rs:5     lib::foo1();
                  ^~~~~~~~~
src/main.rs:5:5: 5:14 help: run `rustc --explain E0425` to see a detailed explanation
error: aborting due to previous error
error: Could not compile `app`.

To learn more, run the command again with --verbose.

```

`cargo run` produces a weird error:

```
$ /data/code/cargo/target/debug/cargo run
   Compiling lib v0.1.0 (file:///data/code/repros/cargo/%232932/lib)
   Compiling lib v0.2.0 (file:///data/code/repros/cargo/%232932/lib-v2)
error: failed to remove /data/code/repros/cargo/#2932/target/debug/deps/lib.0.bc: No such file or directory (os error 2)
error: failed to remove /data/code/repros/cargo/#2932/target/debug/deps/lib.0.o: No such file or directory (os error 2)
error: failed to remove /data/code/repros/cargo/#2932/target/debug/deps/lib.metadata.o: No such file or directory (os error 2)
Build failed, waiting for other jobs to finish...
error: An unknown error occurred

To learn more, run the command again with --verbose
```

**Comment by alexcrichton** [comment]
Hm did this actually show up in practice? It was never intended for two crates of the same name to exist in a DAG that are both path dependencies, I think that'd wreak havoc with the lock file parsing routines!

**Comment by bennofs** [comment]
No, but then at least cargo should give a better error message then.

**Comment by alexcrichton** [comment]
Certainly! I'll tag it as such.

**Comment by kellda** [comment]
I was unable to reproduce with the given project and cargo 1.43.0. Likely this issue is solved ?

**Comment by ehuss** [comment]
Looks like it was fixed around 1.15.
